# Single-cell RNA-ATAC Multiomic Data Analysis â€“ A Tutorial for Single-Cell Biology

In this tutorial, we will use single-cell [RNA-ATAC multiomic data from 18-day-old brain organoids](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-12002), as described in the study [Inferring and perturbing cell fate regulomes in human brain organoids](https://www.nature.com/articles/s41586-022-05279-8#Sec5). We will analyze the scMultiome data in R using `Seurat` and `Signac`.

ðŸ§­ What Youâ€™ll Learn

This tutorial covers the following key parts:
	1.	**Setup**: Understand the experimental design.
	2.	**Load the data:** Explore the RNA and ATAC matrices generated by the 10x Genomics pipeline.
	3.	**Unimodal analysis**: Preprocess RNA and ATAC data separately.
	4.	**Multimodal analysis**: Integrate RNA and ATAC modalities using WNN.
	5.	**Downstream analysis**: Identify DEGs, explore chromatin accessibility around marker genes, and perform motif enrichment analysis.
	6.	**Whatâ€™s next?:** Additional analyses and ideas for taking your project further.

**The tutorial is available in RMarkdown (.Rmd) and HTML formats.**
Use the RMarkdown file to run the analysis locally, or view the HTML version for visual exploration.

Before start this tutorial, make sure you have [R and Rstudio](https://posit.co/download/rstudio-desktop/) installed on your local computer, and have these packages installed:

```
library(EnsDb.Hsapiens.v86) # Annotation database 
library(biovizBase)
library(Seurat)
library(Signac)
library(Matrix)
library(dplyr)
library(tidyr)
library(ggplot2)
library(presto)
library(pheatmap)
library(yaml)
library(tibble)
library(clustree)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggrepel)
library(DESeq2)
```
